dialer.go

```diff
@@ -520,6 +561,7 @@ func (i *instrumentedConn) Write(b []byte) (int, error) {
 // Close delegates to the underlying net.Conn interface and reports the close
 // to the provided closeFunc only when Close returns no error.
 func (i *instrumentedConn) Close() error {
+	i.closed = true
```

Comment:
This isn't thread safe since we have a read happening below on this value.

Reply:
Instead of tracking closed (which requires more locking), could we just always close the connection and ignore double closed errors?